packages_v5/example.candy

```diff
@@ -368,11 +411,142 @@ fun replace[T](list: List[T], index: Int, item: T) List[T] {
+  switch list.isEmpty() {
+    true => none[T](),
+    false => some(list.reduceHelper(1, list.get(0).unwrap(), combine)),
+  }
+}
+fun reduceHelper[T](list: List[T], index: Int, current: T, combine: (T, T) T) T {
+  switch index.isLessThan(list.length()) {
+    true => list.reduceHelper(index.add(1), combine(current, list.get(index).unwrap()), combine),
+    false => current,
+  }
+}
+
+fun all[T](list: List[T], predicate: (T) Bool) Bool {
+  list.allHelper(predicate, 0)
+}
+fun allHelper[T](list: List[T], predicate: (T) Bool, index: Int) Bool {
+  switch index.isLessThan(list.length()) {
+    true => switch predicate(list.get(index).unwrap()) {
+      true => list.allHelper(predicate, index.add(1)),
+      false => false,
+    },
+    false => true,
+  }
+}
+fun any[T](list: List[T], predicate: (T) Bool) Bool {
+  list.all((it: T) { predicate(it).not() }).not()
+}
+
+fun isStrictlyAscending[T: Compare](list: List[T]) Bool {
+  list.isSortedBy((a: T, b: T) { a.compareTo(b).equals(Ordering.less()) })
+}
+fun isStrictlyDescending[T: Compare](list: List[T]) Bool {
+  list.isSortedBy((a: T, b: T) { a.compareTo(b).equals(Ordering.greater()) })
+}
+fun isSortedBy[T](list: List[T], checkPair: (T, T) Bool) Bool {
+  list.pairs().all((pair: Pair[T, T]) { checkPair(pair.first, pair.second) })
+}
+
+fun map[T, R](list: List[T], transform: (T) R) List[R] {
+  list.fold(listOf[R](), (result: List[R], item: T) {
+    result.append(transform(item))
+  })
+}
+fun filter[T](list: List[T], predicate: (T) Bool) List[T] {
+  list.fold(listOf[T](), (result: List[T], item: T) {
+    switch predicate(item) {
+      true => result.append(item),
+      false => result,
+    }
+  })
+}
+fun windows[T](list: List[T], windowLength: Int) List[List[T]] {
+  # Returns a list over all contiguous windows of length `windowLength`.
+  #
+  # The windows overlap. If the `list` is shorter than `windowLength`, the
+  # resulting list is empty.
+  needs(windowLength.isPositive())
+
+  list.windowsHelper(windowLength, listOf[List[T]]())
+}
+fun windowsHelper[T](list: List[T], windowLength: Int, resultSoFar: List[List[T]]) List[List[T]] {
+  let index = resultSoFar.length()
+  switch index.isAtMost(list.length().subtract(windowLength)) {
+    true => list.windowsHelper(
+      windowLength,
+      resultSoFar.append(list.getRange(index, index.add(windowLength))),
+    ),
+    false => resultSoFar,
+  }
+}
+fun pairs[T](list: List[T]) List[Pair[T, T]] {
+  # Returns a list over all contiguous pairs of items.
+  #
+  # The pairs overlap. If the `list` is shorter than two, the resulting list is
+  # empty.
+  list.windows(2).map((window: List[T]) { window.toPair().unwrap() })
+}
+impl[T: ToText] List[T]: ToText {
+  fun toText(self: List[T]) Text {
+    let items = self.map((item: T) { item.toText() })
+      .reduce((result: Text, item: Text) { "{result}, {item}" })
+      .unwrapOr("")
+    "[{items}]"
+  }
+}
+
+struct Pair[T0, T1] {
+  first: T0,
+  second: T1,
+}
```

Comment:
In Martinaise, I use A and B as types and a and b as field names. Opinions?

(Although, in my experience, that pattern leads to compiler errors when you try to introduce a type named C as a compiler stage.)

Reply:
I think I'll leave it as a generic `T` for now, but use different type parameter names if they correspond to something specific (like how we use a lifetime `'h` for referencing the HIR)